client/src/main/kotlin/io/hstream/impl/BufferedProducerKtImpl.kt

```diff
@@ -0,0 +1,130 @@
+        if (closed) {
+            throw HStreamDBClientException("BufferedProducer is closed")
+        }
+        return addToBuffer(hStreamRecord)
+    }
+
+    override fun flush() {
+        futureForIO { lock.lock() }.join()
+        try {
+            if (recordBuffer.isEmpty()) {
+                return
+            } else {
+                val recordBufferCount = recordBuffer.size
+                logger.info("ready to flush recordBuffer, current buffer size is [{}]", recordBufferCount)
+                super.writeHStreamRecords(recordBuffer)
+                    // WARNING: Do not explicitly mark the type of 'recordIds'!
+                    //          The first argument of handle is of type 'List<RecordId>!'.
+                    //          If it is explicitly marked as 'List<RecordId>', a producer
+                    //          will throw an exception but can not be handled because of
+                    //          inconsistent type when it exhausts its retry times. This
+                    //          causes the whole program to be stuck forever.
+                    .handle<Any?> { recordIds, exception: Throwable? ->
+                        if (exception == null) {
+                            for (i in recordIds.indices) {
+                                futures[i].complete(recordIds[i])
+                            }
+                        } else {
+                            for (i in futures.indices) {
+                                futures[i].completeExceptionally(exception)
+                            }
+                        }
+                        null
+                    }
+                    .join()
+                recordBuffer.clear()
+                futures.clear()
+                bytesSize = 0
+                logger.info("flush the record buffer successfully")
+                semaphore.release(recordBufferCount)
+            }
+        } finally {
+            lock.unlock()
+        }
+    }
+
+    private fun addToBuffer(hStreamRecord: HStreamRecord): CompletableFuture<RecordId> {
+        try {
+            semaphore.acquire()
+        } catch (e: InterruptedException) {
+            throw HStreamDBClientException(e)
+        }
+        futureForIO { lock.lock() }.join()
+
+        var needFlush = false
+        val completableFuture = CompletableFuture<RecordId>()
+        try {
+            recordBuffer.add(hStreamRecord)
+            futures.add(completableFuture)
+            bytesSize += hStreamRecord.payload.size()
+            if (recordBuffer.size == recordCountLimit) {
+                needFlush = true
+            } else if (maxBytesSize >= 0 && bytesSize > maxBytesSize) {
+                needFlush = true
+            }
+        } finally {
+            lock.unlock()
+        }
+        if (needFlush) {
+            flush()
+        }
+        return completableFuture
+    }
+
+    private fun runTimer() {
+        timerScope.launch {
```

Comment:
this coroutine will be leaked and we should cancel it after the bufferedProducer was closed.